src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs

```diff
@@ -300,13 +293,49 @@ public virtual async Task<T> GetConfigurationAsync(CancellationToken cancel)
                     fetchMetadataFailure));
         }
 
+        private void EnsureBackgroundTaskIsRunning()
+        {
+            if (_updateMetadataTask == null || _updateMetadataTask.Status != TaskStatus.Running)
+                _updateMetadataTask = Task.Run(UpdateCurrentConfigurationUsingSignals, CancellationToken.None);
+        }
+
+        private void TelemetryForUpdate()
+        {
+            var updateMode = _refreshRequested ? TelemetryConstants.Protocols.Manual : TelemetryConstants.Protocols.Automatic;
+
+            if (_refreshRequested)
+                _refreshRequested = false;
+
+            try
+            {
+                TelemetryClient.IncrementConfigurationRefreshRequestCounter(
+                    MetadataAddress,
+                    updateMode);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch
+            { }
+#pragma warning restore CA1031 // Do not catch general exception types
+        }
+
+        private void UpdateCurrentConfigurationUsingSignals()
+        {
+            while (true)
+            {
+                _updateMetadata.WaitOne();
+                UpdateCurrentConfiguration();
+                _onBackgroundTaskFinish?.Invoke();
```

Comment:
This is a bit unfortunate, but the motivation is understandable.

Reply:
I agree, I don't like have code here only for testing. But this allows us to have stable tests, without more invasive changes.